api: add listener_address_restriction_hint in Node to hint known listening ports on the node

sanjaypujare opened the pull request:
Description: Added an optional field in `Node` as a generic hint to the management server about known listening ports on the node

Risk Level: Low
Testing: manually tested by compiling protos into generated Java code and using the code in a Java project
Docs Changes: N/A
Release Notes: N/A

  Signed-off-by: Sanjay Pujare <sanjaypujare@users.noreply.github.com>

sanjaypujare commented:
@htuch pls review

ejona86 commented:
To be more clear about the use-case: This is for cases where the server has already decided which ports it will listen on and needs configuration for those ports.

sanjaypujare commented:
Looked into the `envoy-linux (format)` failure:
```
-import "envoy/config/core/v3/base.proto";
+import "envoy/api/v3/core/socket_option.proto";
```

For some reason `socket_option.proto` was not properly placed in the v3 hierarchy.  It should go into `envoy/config/core/v3`. I think I need to update some file to include socket_option.proto there.

alyssawilk commented:
@htuch @lizan given this is API work one of you up for taking this?

markdroth commented:
This was suggested by @htuch, so I assume he'll review.

sanjaypujare commented:
@htuch Just FYI I did try to test the changes on my Mac but there are infra/tooling problems I am unable to debug. Pls let me know if the failures are related to my changes.

htuch commented on api/envoy/api/v2/core/base.proto:
```diff
@@ -182,6 +184,12 @@ message Node {
+  // for filtering :ref:`listeners <config_listeners>` to be returned. For example,
+  // if there is a listener bound to port 80, the list can optionally contain the
+  // SocketAddress `(0.0.0.0,80)`. The field is optional and just a hint.
+  repeated Address listener_address_restriction_hint = 11;
```
This PR looks good, just trying to see if we can maybe name this a bit more descriptively (I know I was responsible for the original suggestion :P)

How about `listener_bound_addresses`, `listening_addresses`, `listener_available_addresses`?

sanjaypujare replied:
I'll pick `listening_addresses` as the shortest of them all.

sanjaypujare commented:
> LGTM, but you need to run `fix_format` (to forward propagate into v3) ...

So I ran 
```
./ci/run_envoy_docker.sh './ci/do_ci.sh fix_format'
```
And that finished successfully and generated following changes:
```
	modified:   api/BUILD
	modified:   api/envoy/api/v2/core/base.proto
	modified:   api/envoy/api/v2/core/socket_option.proto
	modified:   api/envoy/api/v2/listener.proto
	modified:   api/envoy/config/bootstrap/v2/bootstrap.proto
	modified:   api/envoy/config/bootstrap/v3/BUILD
	modified:   api/envoy/config/bootstrap/v3/bootstrap.proto
	modified:   api/envoy/config/core/v3/BUILD
	modified:   api/envoy/config/core/v3/address.proto
	modified:   api/envoy/config/core/v3/base.proto
	modified:   api/envoy/config/listener/v3/BUILD
	modified:   api/envoy/config/listener/v3/listener.proto
	modified:   generated_api_shadow/envoy/api/v2/core/address.proto
	modified:   generated_api_shadow/envoy/api/v2/core/base.proto
	modified:   generated_api_shadow/envoy/api/v2/listener.proto
	modified:   generated_api_shadow/envoy/config/bootstrap/v2/bootstrap.proto
	modified:   generated_api_shadow/envoy/config/bootstrap/v3/BUILD
	modified:   generated_api_shadow/envoy/config/bootstrap/v3/bootstrap.proto
	modified:   generated_api_shadow/envoy/config/core/v3/BUILD
	modified:   generated_api_shadow/envoy/config/core/v3/address.proto
	modified:   generated_api_shadow/envoy/config/core/v3/base.proto
	modified:   generated_api_shadow/envoy/config/listener/v3/BUILD
	modified:   generated_api_shadow/envoy/config/listener/v3/listener.proto

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	api/envoy/api/v3/
	generated_api_shadow/envoy/api/v2/core/socket_option.proto
	generated_api_shadow/envoy/api/v3/

```

Should I commit all these changes? And more importantly, why is `socket_options`  under `api/envoy/api/v3/core/` instead of `api/envoy/config/core/v3` like other similar files?

htuch commented on api/envoy/api/v2/core/socket_option.proto:
```diff
@@ -0,0 +1,47 @@
+
+package envoy.api.v2.core;
+
+import "validate/validate.proto";
```
You'll need to add option `(udpa.annotations.file_migrate).move_to_package = "envoy.config.core.v3";` (as done in other files in this package) for the correct file movement in v3.
`

sanjaypujare commented:
> @sanjaypujare good catch on the v3 file move:
> 
> 1. Yes, please add all the new shadow files.
> 2. See my comment on how to ensure the new socket options file is in the right place.
>    /wait

@htuch done. PTAL.

htuch commented:
@sanjaypujare I think you ended up colliding with a known master breakage. Can you sync past https://github.com/envoyproxy/envoy/pull/10426, merge master and try again?

You will also need to fix DCO: https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco

sanjaypujare commented:
> @sanjaypujare I think you ended up colliding with a known master breakage. Can you sync past #10426, merge master and try again?
> 
> You will also need to fix DCO: https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#fixing-dco

Done both. PTAL.

htuch commented:
@sanjaypujare see the docs warning:
```
checking consistency... /source/generated/rst/api-v3/config/core/v3/socket_option.proto.rst: WARNING: document isn't included in any toctree
```
I suggest adding to https://github.com/envoyproxy/envoy/blob/master/docs/root/api-v3/common_messages/common_messages.rst  and https://github.com/envoyproxy/envoy/blob/master/docs/root/api-v2/common_messages/common_messages.rst as appropriate.

sanjaypujare commented:
@htuch v3 common_messages.rst updated (v2 was already done).

lizan commented on api/envoy/api/v2/core/base.proto:
```diff
@@ -182,6 +185,12 @@ message Node {
+  // for filtering :ref:`listeners <config_listeners>` to be returned. For example,
+  // if there is a listener bound to port 80, the list can optionally contain the
+  // SocketAddress `(0.0.0.0,80)`. The field is optional and just a hint.
+  repeated Address listening_addresses = 11;
```
This is not implemented yet in this PR? so hide from doc? or do you plan to implement them?

sanjaypujare replied:
Not sure about what is "not implemented". We are planning to use this new field in our xDS client (proxyless gRPC) as a hint to the xDS server for particular listener(s). I don't see any reason hide it in the docs. 

Also note that we had some offline discussion with @htuch about this and this was the agreed upon approach.

htuch replied:
@lizan the user specifics it in bootstrap, and this then gets automagically dumped into each xDS stream (we merge `Node` messages).

lizan replied:
@htuch ah ok I see how it works now. Thanks.

sanjaypujare commented:
> LGTM

@htuch thanks! Once this is merged the downstream design/implementation activities will continue.
